lib/web_ui/lib/src/engine/layers.dart

```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
     if (other is! PlatformViewStyling) {
       return false;
     }
-    return (position == other.position) && (opacity == other.opacity);
+    return (position == other.position) && (opacity == other.opacity) && (clip == other.clip);
   }
 
   @override
   int get hashCode {
-    return Object.hash(position, opacity);
+    return Object.hash(position, opacity, clip);
   }
 }
 
+sealed class PlatformViewClip {
+  PlatformViewClip positioned(PlatformViewPosition position);
+
+  ui.Rect get boundingRect;
+
+  bool covers(ui.Rect rect);
+
+  ScenePath get asPath;
+
+  static PlatformViewClip combine(PlatformViewClip outer, PlatformViewClip inner) {
+    if (outer is PlatformViewNoClip) {
+      return inner;
+    }
+    if (inner is PlatformViewNoClip) {
+      return outer;
+    }
+
+    if (outer.covers(inner.boundingRect)) {
+      return outer;
+    }
+
+    if (inner.covers(outer.boundingRect)) {
+      return inner;
+    }
+
+    final ScenePath path = ui.Path() as ScenePath;
+    path.addPath(outer.asPath, ui.Offset.zero);
+    path.addPath(inner.asPath, ui.Offset.zero);
+    return PlatformViewPathClip(path);
+  }
+}
+
+class PlatformViewNoClip implements PlatformViewClip {
+  const PlatformViewNoClip();
+
+  @override
+  PlatformViewClip positioned(PlatformViewPosition positioned) {
+    return this;
+  }
+
+  @override
+  ScenePath get asPath => ui.Path() as ScenePath;
+
+  @override
+  ui.Rect get boundingRect => ui.Rect.zero;
+
+  @override
+  bool covers(ui.Rect rect) {
+    return false;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other) || (other.runtimeType == runtimeType)) {
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  int get hashCode => runtimeType.hashCode;
```

Comment:
I think I actually tried this, but `PlatformViewNoClip.hashCode` doesn't actually work because the language will attempt to call a static method called `hashCode` which doesn't exist. I also feel like hashing the runtime type is better than assigning an arbitrary value, because some other equally cute person might return `42` for their class which is also a singleton type, and if they end up in a very generic (i.e. `<Object>`) hashed collection together they will have a hash collision. Is this class likely to be put in a hashed collection with other random singleton classes? No. In practice it isn't even put in a hashed collection at all, but the language wants me to implement this if I implement `operator==`, so this is all theoretical and probably will never matter anyway.